proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+#### `fiber.resume` Resume a suspended fiber
+
+The `fiber.resume` instruction takes a fiber as argument, together with an `event` description&mdash;consisting of an event tag and appropriate values, and resumes its execution.
+
+The `fiber.resume` instruction takes a `suspended` fiber, together with any descendant fibers that were suspended along with it, and resumes its execution. The event is used to encode how the resumed fiber should react: for example, whether the fiber's requested information is available, or whether the fiber should enter into cancelation mode.
+
+A `fiber.resume` instruction may continue in one of two ways: if the resumed fiber returned normally, then execution continues with the next instruction. The values on the value stack will match the return signature of the fiber's function&mdash;and will match the signature of the `fiber` reference that was resumed.
+
+If the resumed fiber suspended itself, then the event tag associated with that `fiber.suspend` instruction is used to determine which of the available `event` blocks should be entered as part of the switch. The 'nearest' `event` block whose tag is equal to the supplied event is entered. If there is no appropriate `event` block in the execution scope of the fiber being resumed, then the engine _traps_.
+
+#### `fiber.switch` Switch to a different fiber
+
+The `fiber.switch` instruction is a combination of a `fiber.suspend` and a `fiber.resume` to an identified fiber. This instruction is useful for circumstances where the suspending fiber knows which other fiber should be resumed.
+
+The `fiber.switch` instruction has three arguments: the identity of the fiber being suspended, the identity of the fiber being resumed and the signaling event.
+
+Although it may be viewed as being a combination of the two instructions, there is an important distinction also: the signaling event. Under the common hierarchical organization, a suspending fiber does not know which fiber will be resumed. This means that the signaling event has to be of a form that the fiber's manager is ready to process. However, with a `fiber.switch` instruction, the fiber's manager is not informed of the switch and does not need to understand the signaling event.
+
+This, in turn, means that a fiber manager may be relieved of the burden of communicating between fibers. I.e., `fiber.switch` supports a symmetric coroutining pattern. However, precisely because the fiber's manager is not made aware of the switch between fibers, it must also be the case that this does not _matter_; in effect, the fiber manager may not directly be aware of any of the fibers that it is managing.  
+
+#### `fiber.retire` Retire a fiber
```

Comment:
This seems like it's only useful for exceptions, since fibers can always return nonexceptionally without any event involved. What if we made propagating an exception out of a fiber behave the same as doing a `fiber.retire` with the same tag as the exception? Then we might not need a separate `fiber.retire` instruction.

(Relatedly, is there any reason not to reuse try-catch rather than introducing `handle-event`? They seem awfully similar, and reusing try-catch would be the same as throwing an exception when receiving an event for which there is no handler, which does not seems necessarily like a worse choice than trapping. It also seems like it would be useful to have a handle-all analogous to catch-all.)

Reply:
exceptions have a policy attached of automatic propagation of uncaught exceptions. This is not something that we want. Nor is there any equivalent of catchall, delegate etc in the fiber model. But, otherwise, there is some similarity. 
It is true that fiber.retire would be useful for exceptions; it is not the only use. It can also be useful when escaping out of deep generators.

Reply:
Why is `fiber.retire` more useful than normal return for escaping deep generators?

It's not clear to me that the policy of trapping on unhandled events is obviously better than throwing an exception on unhandled events. I also certainly anticipate a need to recover gracefully from unexpected events, and in the fullness of time I think we would want exception-throwing versions of these primitives even if we start out with trapping primitives. (Having throwing versions of currently-trapping instructions is a top request in the GC proposal, for instance.)

Reply:
That may be. At the moment, it is less risky and involves less design to trap.